src/freenet/support/TimeUtil.java

```diff
@@ -135,77 +138,49 @@ public static String formatTime(long timeInterval, int maxTerms) {
     }
 
     public static long toMillis(String timeInterval) {
-        byte sign = 1;
-        if (timeInterval.contains("-")) {
-            sign = -1;
-            timeInterval = timeInterval.substring(1);
+        Matcher matcher = TIME_INTERVAL_PATTERN.matcher(timeInterval);
+        if (!matcher.matches()) {
+            throw new NumberFormatException("Unknown format: " + timeInterval);
         }
 
-        String[] terms = timeInterval.split("(?<=[a-z])");
-
+        String group;
         long millis = 0;
-        for (String term : terms) {
-            if (term.isEmpty()) continue;
-
-            char measure = term.charAt(term.length() - 1);
-            switch(measure){
-                case 'w':
-                    millis += 7 * MILLISECONDS.convert(Long.parseLong(term.substring(0, term.length() - 1)), DAYS);
-                    break;
-                case 'd':
-                    millis += MILLISECONDS.convert(Short.parseShort(term.substring(0, term.length() - 1)), DAYS);
-                    break;
-                case 'h':
-                    millis += MILLISECONDS.convert(Short.parseShort(term.substring(0, term.length() - 1)), HOURS);
-                    break;
-                case 'm':
-                    millis += MILLISECONDS.convert(Short.parseShort(term.substring(0, term.length() - 1)), MINUTES);
-                    break;
-                case 's':
-                    if (term.contains(".")) {
-                        millis += Integer.parseInt(term.replaceAll("[a-z.]", ""));
-                    } else {
-                        millis += MILLISECONDS.convert(Short.parseShort(term.substring(0, term.length() - 1)), SECONDS);
-                    }
-                    break;
-                default:
-                    throw new NumberFormatException("Unknown format: " + (sign > 0 ? "" : "-") + timeInterval);
-            }
+        if ((group = matcher.group(1)) != null) { // weeks
+            millis += DAYS.toMillis(7 * Long.parseLong(group));
+        }
+        if ((group = matcher.group(2)) != null) { // days
```

Comment:
Better with else if ?

Reply:
That wouldn't work, all `if` clauses can (and should) match independently as they operate on different parts of the input.